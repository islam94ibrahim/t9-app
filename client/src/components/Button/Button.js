import React from 'react';

import styles from './Button.module.css';

const Button = (props) => {
  const classes = [styles.Button];

  if (props.active) {
    classes.push(styles.Active);
  }

  return (
    <button
      className={classes.join(' ')}
      disabled={props.disabled}
      onClick={() => props.clicked(props.value)}
    >
      {props.value}
      <span>{props.letters}</span>
    </button>
  );
};

export default Button;
