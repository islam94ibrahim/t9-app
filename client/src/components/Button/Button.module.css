.Button {
  height: 50px;
  width: 75px;
}

.Button span {
  display: block;
  margin-top: 5px;
  font-size: 10px;
}

.Active {
  background-color: #4caf50;
  border: 1px solid #ccc;
  border-radius: 4px;
}
