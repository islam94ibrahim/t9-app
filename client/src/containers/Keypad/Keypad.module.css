.Keypad h1 {
  text-align: center;
  margin-bottom: 30px;
}

.Keypad table {
  margin: auto;
}

.Keypad input {
  width: 225px;
  height: 25px;
  margin-left: 2px;
  text-align: right;
}
