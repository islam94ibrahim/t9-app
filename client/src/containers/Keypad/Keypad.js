import React, { useState, useEffect, useCallback } from 'react';
import axios from 'axios';

import styles from './Keypad.module.css';
import Button from '../../components/Button/Button';
import Board from '../../components/Board/Board';

const KEYPAD_BUTTONS = [
  { value: 1, letters: '. , !', disabled: true },
  { value: 2, letters: 'a b c' },
  { value: 3, letters: 'd e f' },
  { value: 4, letters: 'g h i' },
  { value: 5, letters: 'j k l' },
  { value: 6, letters: 'm n o' },
  { value: 7, letters: 'p q r s' },
  { value: 8, letters: 't u v' },
  { value: 9, letters: 'w x y z' },
  { value: '*', disabled: true },
  { value: 0, disabled: true },
  { value: '#', disabled: true },
  { value: '#', disabled: true },
];

const Keypad = () => {
  const [numbers, setNumbers] = useState('');
  const [onlyRealWords, setOnlyRealWords] = useState(false);
  const [words, setWords] = useState([]);

  const requestPrediction = useCallback(() => {
    const body = { numbers, onlyRealWords };

    axios
      .post('/predict', body)
      .then(({ data }) => {
        setWords(data.words.join(', '));
      })
      .catch((error) => {
        setWords(error.message);
      });
  }, [numbers, onlyRealWords]);

  useEffect(() => {
    if (numbers !== '') requestPrediction();
  }, [numbers, requestPrediction]);

  const buttonHandler = (buttonValue) => {
    setNumbers(numbers + buttonValue);
  };

  const deleteButtonHandler = () => {
    setNumbers(numbers.slice(0, -1));
  };

  return (
    <div className={styles.Keypad}>
      <h1>T9 application</h1>
      <table>
        <tbody>
          <tr>
            <td colSpan="3">
              <input type="text" value={numbers} readOnly />
            </td>
          </tr>
          <tr>
            <td colSpan="2">
              <Button
                value="R"
                letters="Real words"
                active={onlyRealWords}
                clicked={() => setOnlyRealWords(!onlyRealWords)}
              />
            </td>
            <td>
              <Button
                value="X"
                letters="DELETE"
                clicked={deleteButtonHandler}
              />
            </td>
          </tr>
          <tr>
            {KEYPAD_BUTTONS.slice(0, 3).map((button, index) => (
              <td key={index}>
                <Button {...button} clicked={buttonHandler} />
              </td>
            ))}
          </tr>
          <tr>
            {KEYPAD_BUTTONS.slice(3, 6).map((button, index) => (
              <td key={index}>
                <Button {...button} clicked={buttonHandler} />
              </td>
            ))}
          </tr>
          <tr>
            {KEYPAD_BUTTONS.slice(6, 9).map((button, index) => (
              <td key={index}>
                <Button {...button} clicked={buttonHandler} />
              </td>
            ))}
          </tr>
          <tr>
            {KEYPAD_BUTTONS.slice(9, 12).map((button, index) => (
              <td key={index}>
                <Button {...button} clicked={buttonHandler} />
              </td>
            ))}
          </tr>
        </tbody>
      </table>
      <Board>{words}</Board>
    </div>
  );
};

export default Keypad;
