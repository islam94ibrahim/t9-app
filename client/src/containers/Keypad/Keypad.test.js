import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Keypad from './Keypad';
import Button from '../../components/Button/Button';

configure({ adapter: new Adapter() });

describe('<Keypad />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Keypad />);
  });

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders 14 buttons represnting mobile keypad', () => {
    expect(wrapper.find(Button)).toHaveLength(14);
  });
});
