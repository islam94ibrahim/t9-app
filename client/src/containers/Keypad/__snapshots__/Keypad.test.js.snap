// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Keypad /> renders correctly 1`] = `"<div class=\\"Keypad\\"><h1>T9 application</h1><table><tbody><tr><td colSpan=\\"3\\"><input type=\\"text\\" value=\\"\\" readonly=\\"\\"/></td></tr><tr><td colSpan=\\"2\\"><button class=\\"Button\\">R<span>Real words</span></button></td><td><button class=\\"Button\\">X<span>DELETE</span></button></td></tr><tr><td><button class=\\"Button\\" disabled=\\"\\">1<span>. , !</span></button></td><td><button class=\\"Button\\">2<span>a b c</span></button></td><td><button class=\\"Button\\">3<span>d e f</span></button></td></tr><tr><td><button class=\\"Button\\">4<span>g h i</span></button></td><td><button class=\\"Button\\">5<span>j k l</span></button></td><td><button class=\\"Button\\">6<span>m n o</span></button></td></tr><tr><td><button class=\\"Button\\">7<span>p q r s</span></button></td><td><button class=\\"Button\\">8<span>t u v</span></button></td><td><button class=\\"Button\\">9<span>w x y z</span></button></td></tr><tr><td><button class=\\"Button\\" disabled=\\"\\">*<span></span></button></td><td><button class=\\"Button\\" disabled=\\"\\">0<span></span></button></td><td><button class=\\"Button\\" disabled=\\"\\">#<span></span></button></td></tr></tbody></table><div class=\\"Board\\"><h2>Words</h2><div></div></div></div>"`;
